exporter/loadbalancingexporter/exporter.go

```diff
@@ -16,42 +16,208 @@ package loadbalancingexporter
+		ctx := context.Background()
+
+		// add the missing exporters first
+		e.addMissingExporters(ctx, resolved)
+		e.removeExtraExporters(ctx, resolved)
+	}
+}
+
+func (e *exporterImp) addMissingExporters(ctx context.Context, endpoints []string) {
+	for _, endpoint := range endpoints {
+		endpoint = endpointWithPort(endpoint)
+
+		if _, exists := e.exporters[endpoint]; !exists {
+			cfg := e.exporterFactory.CreateDefaultConfig()
+			oCfg := cfg.(*otlpexporter.Config)
+
+			// is this wrong? are we overriding default fields, or is the config unmarshaller taking care of it?
+			oCfg.ExporterSettings = e.config.Protocol.OTLP.ExporterSettings
+			oCfg.TimeoutSettings = e.config.Protocol.OTLP.TimeoutSettings
+			oCfg.QueueSettings = e.config.Protocol.OTLP.QueueSettings
+			oCfg.RetrySettings = e.config.Protocol.OTLP.RetrySettings
+			oCfg.GRPCClientSettings = e.config.Protocol.OTLP.GRPCClientSettings
+
+			oCfg.Endpoint = endpoint
+			exp, err := e.exporterFactory.CreateTracesExporter(ctx, e.templateCreateParams, cfg)
+			if err != nil {
+				e.logger.Warn("failed to create new trace exporter for endpoint", zap.String("endpoint", endpoint))
+				continue
+			}
+			e.exporters[endpoint] = exp
+		}
+	}
+}
+
+func (e *exporterImp) removeExtraExporters(ctx context.Context, endpoints []string) {
+	for existing := range e.exporters {
+		if !endpointFound(existing, endpoints) {
+			e.exporters[existing].Shutdown(ctx)
+			delete(e.exporters, existing)
+		}
+	}
+}
+
+func endpointFound(endpoint string, endpoints []string) bool {
+	for _, candidate := range endpoints {
+		if candidate == endpoint {
+			return true
+		}
+	}
+
+	return false
+}
+
 func (e *exporterImp) Shutdown(context.Context) error {
+	e.stopped = true
+	e.shutdownWg.Wait()
 	return nil
 }
 
 func (e *exporterImp) ConsumeTraces(ctx context.Context, td pdata.Traces) error {
-	return nil
+	e.updateLock.RLock()
+	defer e.updateLock.RUnlock()
+
+	traceID := traceIDFromTraces(td)
+	endpoint := e.ring.endpointFor(traceID)
+	exp, found := e.exporters[endpoint]
+	if !found {
+		// something is really wrong... how come we couldn't find the exporter??
+		return fmt.Errorf("couldn't find the exporter for the endpoint %q", endpoint)
+	}
+
+	start := time.Now()
+	err := exp.ConsumeTraces(ctx, td)
+	duration := time.Since(start)
+	ctx, _ = tag.New(ctx, tag.Upsert(tag.MustNewKey("endpoint"), endpoint))
+
+	if err == nil {
+		sCtx, _ := tag.New(ctx, tag.Upsert(tag.MustNewKey("success"), "true"))
+		stats.Record(sCtx, mBackendLatency.M(duration.Milliseconds()))
+	} else {
+		fCtx, _ := tag.New(ctx, tag.Upsert(tag.MustNewKey("success"), "false"))
+		stats.Record(fCtx, mBackendLatency.M(duration.Milliseconds()))
+	}
+
+	return err
 }
 
 func (e *exporterImp) GetCapabilities() component.ProcessorCapabilities {
 	return component.ProcessorCapabilities{MutatesConsumedData: false}
 }
+
+func traceIDFromTraces(td pdata.Traces) pdata.TraceID {
+	// is this safe? can a trace be empty and not contain a single span?
```

Comment:
Does the exporter rely on groupByTraceProcessor? I would expect that to only export non-empty traces and it could be considered a bug there. That being said, I don't think we should panic here if something's up, can we filter empty traces here?

Reply:
Yes, it does rely on the `groupByTrace` at this moment, but when I built this, I thought that we'd be having another similar processor that doesn't hold traces in memory, and would just reorder the incoming batches. This last processor later became a library, so, I should probably update this code to use that instead.

Reply:
Got it - we can do a refactor with using that library in a followup, it'd be good to go ahead and tweak this code to remove the chances of a panic if there happen to be no spans.